pkg/http/api.go

```diff
@@ -2020,6 +2025,218 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
+			entries[i].Metadata = populateMetadataForBulkPublishEntry(metadata, entry.Metadata)
+		}
+		if entry.EntryID == "" {
+			msg := NewErrorResponse("ERR_PUBSUB_EVENTS_SER",
+				fmt.Sprintf(messages.ErrPubsubMarshal, topic, pubsubName, "error: entryID is not present for entry"))
+			respond(reqCtx, withError(fasthttp.StatusBadRequest, msg))
+			log.Debug(msg)
+
+			return
+		}
+		entries[i].EntryID = entry.EntryID
+	}
+
+	spanMap := map[int]trace.Span{}
+	// closeChildSpans method is called on every respond() call in all return paths in the following block of code.
+	closeChildSpans := func(ctx *fasthttp.RequestCtx) {
+		for _, span := range spanMap {
+			diag.UpdateSpanStatusFromHTTPStatus(span, ctx.Response.StatusCode())
+			span.End()
+		}
+	}
+	if !rawPayload {
+		for i := range entries {
+			// For multiple events in a single bulk call traceParent is different for each event.
+			childSpan := diag.StartProducerSpanChildFromParent(reqCtx, span)
+			// Populate W3C traceparent to cloudevent envelope
+			corID := diag.SpanContextToW3CString(childSpan.SpanContext())
+			spanMap[i] = childSpan
+
+			var envelope map[string]interface{}
+
+			envelope, err = runtimePubsub.NewCloudEvent(&runtimePubsub.CloudEvent{
+				ID:              a.id,
+				Topic:           topic,
+				DataContentType: entries[i].ContentType,
+				Data:            entries[i].Event,
+				TraceID:         corID,
+				TraceState:      traceState,
+				Pubsub:          pubsubName,
+			})
+			if err != nil {
+				msg := NewErrorResponse("ERR_PUBSUB_CLOUD_EVENTS_SER",
+					fmt.Sprintf(messages.ErrPubsubCloudEventCreation, err.Error()))
+				respond(reqCtx, withError(fasthttp.StatusInternalServerError, msg), closeChildSpans)
+				log.Debug(msg)
+
+				return
+			}
+
+			features := thepubsub.Features()
+
+			pubsub.ApplyMetadata(envelope, features, entries[i].Metadata)
+
+			entries[i].Event, err = json.Marshal(envelope)
+			if err != nil {
+				msg := NewErrorResponse("ERR_PUBSUB_CLOUD_EVENTS_SER",
+					fmt.Sprintf(messages.ErrPubsubCloudEventsSer, topic, pubsubName, err.Error()))
+				respond(reqCtx, withError(fasthttp.StatusInternalServerError, msg), closeChildSpans)
+				log.Debug(msg)
+
+				return
+			}
+		}
+	}
+
+	req := pubsub.BulkPublishRequest{
+		PubsubName: pubsubName,
+		Topic:      topic,
+		Entries:    entries,
+		Metadata:   metadata,
+	}
+
+	start := time.Now()
+	res, err := a.pubsubAdapter.BulkPublish(&req)
+	elapsed := diag.ElapsedSince(start)
+
+	diag.DefaultComponentMonitoring.BulkPubsubEgressEvent(context.Background(), pubsubName, topic, err == nil, elapsed)
+
+	bulkRes := BulkPublishResponse{}
+
+	if len(res.Statuses) != 0 {
+		bulkRes.Statuses = make([]BulkPublishResponseEntry, len(res.Statuses))
+		for i, r := range res.Statuses {
+			bulkRes.Statuses[i].EntryID = r.EntryID
+			if r.Error != nil {
+				bulkRes.Statuses[i].Error = r.Error.Error()
+			}
+			bulkRes.Statuses[i].Status = string(r.Status)
+		}
+	}
+
+	if err != nil {
```

Comment:
In case of error also an empty result is always returned... The plain struct is returned and not a pointer to it... So it will always be empty or have some data...